pdns/recursordist/test-recursorcache_cc.cc

```diff
@@ -892,4 +892,233 @@ BOOST_AUTO_TEST_CASE(test_RecursorCache_Wipe)
+    dr1.d_name = power;
+    dr1.d_type = QType::A;
+    dr1.d_class = QClass::IN;
+    dr1.d_content = std::make_shared<ARecordContent>(dr1Content);
+    dr1.d_ttl = static_cast<uint32_t>(ttd);
+    dr1.d_place = DNSResourceRecord::ANSWER;
+    std::vector<DNSRecord> rset1;
+    rset1.push_back(dr1);
+
+    DNSRecord dr2;
+    ComboAddress dr2Content("192.0.2.42");
+    dr2.d_name = power;
+    dr2.d_type = QType::A;
+    dr2.d_class = QClass::IN;
+    dr2.d_content = std::make_shared<ARecordContent>(dr2Content);
+    dr2.d_ttl = static_cast<uint32_t>(ttd);
+    dr2.d_place = DNSResourceRecord::ANSWER;
+    std::vector<DNSRecord> rset2;
+    rset2.push_back(dr2);
+
+    DNSRecord dr3;
+    ComboAddress dr3Content("192.0.2.43");
+    dr3.d_name = power;
+    dr3.d_type = QType::A;
+    dr3.d_class = QClass::IN;
+    dr3.d_content = std::make_shared<ARecordContent>(dr3Content);
+    dr3.d_ttl = static_cast<uint32_t>(ttd);
+    dr3.d_place = DNSResourceRecord::ANSWER;
+    std::vector<DNSRecord> rset3;
+    rset3.push_back(dr3);
+
+    // insert a tagged entry
+    MRC.replace(now, power, QType(QType::A), rset1, signatures, authRecords, true, boost::optional<Netmask>("192.0.2.0/24"), string("mytag"));
+    BOOST_CHECK_EQUAL(MRC.size(), 1U);
+
+    // tagged specific should be returned for a matching tag
+    BOOST_CHECK_EQUAL(MRC.get(now, power, QType(QType::A), false, &retrieved, ComboAddress("192.0.2.2"), string("mytag")), (ttd - now));
+    BOOST_REQUIRE_EQUAL(retrieved.size(), 1U);
+    BOOST_CHECK_EQUAL(getRR<ARecordContent>(retrieved.at(0))->getCA().toString(), dr1Content.toString());
+
+    // tag specific should not be returned for a different tag
+    BOOST_CHECK_LT(MRC.get(now, power, QType(QType::A), false, &retrieved, ComboAddress("192.0.2.2"), string("othertag")), 0);
+    BOOST_CHECK_EQUAL(retrieved.size(), 0U);
+
+    // insert a new  entry without tag
+    MRC.replace(now, power, QType(QType::A), rset2, signatures, authRecords, true, boost::optional<Netmask>("192.0.3.0/24"), boost::none);
+    BOOST_CHECK_EQUAL(MRC.size(), 2U);
+
+    // tagged specific should be returned for a matching tag
+    BOOST_CHECK_EQUAL(MRC.get(now, power, QType(QType::A), false, &retrieved, ComboAddress("192.0.2.2"), string("mytag")), (ttd - now));
+    BOOST_REQUIRE_EQUAL(retrieved.size(), 1U);
+    BOOST_CHECK_EQUAL(getRR<ARecordContent>(retrieved.at(0))->getCA().toString(), dr1Content.toString());
+
+    // if no tag given nothing should be retrieved if address doesn't match
+    BOOST_CHECK_LT(MRC.get(now, power, QType(QType::A), false, &retrieved, ComboAddress("127.0.0.1"), boost::none), 0);
+    BOOST_REQUIRE_EQUAL(retrieved.size(), 0U);
+
+    // if no tag given and no-non-tagged entries are available nothing should be retrieved even if address  matches
+    BOOST_CHECK_LT(MRC.get(now, power, QType(QType::A), false, &retrieved, ComboAddress("192.0.2.2"), boost::none), 0);
+    BOOST_REQUIRE_EQUAL(retrieved.size(), 0U);
+
```

Comment:
Perhaps add this test to see if we can retrieve an ECS-specific entry with no routing tag at this point:

```C++
    BOOST_CHECK_EQUAL(MRC.get(now, power, QType(QType::A), false, &retrieved, ComboAddress("192.0.3.2"), boost::none), (ttd -now));
    BOOST_REQUIRE_EQUAL(retrieved.size(), 1U);
    BOOST_CHECK_EQUAL(getRR<ARecordContent>(retrieved.at(0))->getCA().toString(), dr2Content.toString());
```